File: backend/animal-registry/src/main/java/com/animalshelter/animalregistry/service/AnimalService.java
package com.animalshelter.animalregistry.service;

import com.animalshelter.animalregistry.config.UserContext;
import com.animalshelter.animalregistry.dto.*;
import com.animalshelter.animalregistry.exception.ResourceNotFoundException;
import com.animalshelter.animalregistry.messaging.AnimalEventPublisher;
import com.animalshelter.animalregistry.messaging.event.AnimalRegisteredEvent;
import com.animalshelter.animalregistry.messaging.event.AnimalStatusChangedEvent;
import com.animalshelter.animalregistry.model.Animal;
import com.animalshelter.animalregistry.model.AnimalCategory;
import com.animalshelter.animalregistry.model.AnimalStatus;
import com.animalshelter.animalregistry.model.StatusHistoryEntry;
import com.animalshelter.animalregistry.repository.AnimalRepository;
import com.animalshelter.animalregistry.repository.MedicalRecordRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class AnimalService {

    private final AnimalRepository animalRepository;
    private final MedicalRecordRepository medicalRecordRepository;
    private final MongoTemplate mongoTemplate;
    private final ImageStorageService imageStorageService;
    private final UserContext userContext;
    private final AnimalEventPublisher eventPublisher;
    private final AnimalVectorService animalVectorService;
    private final int maxImagesPerAnimal;

    public AnimalService(
            AnimalRepository animalRepository,
            MedicalRecordRepository medicalRecordRepository,
            MongoTemplate mongoTemplate,
            ImageStorageService imageStorageService,
            UserContext userContext,
            AnimalEventPublisher eventPublisher,
            AnimalVectorService animalVectorService,
            @Value("${app.upload.max-images-per-animal}") int maxImagesPerAnimal
    ) {
        this.animalRepository = animalRepository;
        this.medicalRecordRepository = medicalRecordRepository;
        this.mongoTemplate = mongoTemplate;
        this.imageStorageService = imageStorageService;
        this.userContext = userContext;
        this.eventPublisher = eventPublisher;
        this.animalVectorService = animalVectorService;
        this.maxImagesPerAnimal = maxImagesPerAnimal;
    }

    public AnimalResponse createAnimal(CreateAnimalRequest request) {
        if (request.chipId() != null && !request.chipId().isBlank()
                && animalRepository.existsByChipId(request.chipId())) {
            throw new IllegalArgumentException("An animal with chip ID " + request.chipId() + " already exists");
        }

        LocalDateTime now = LocalDateTime.now();

        StatusHistoryEntry initialStatus = StatusHistoryEntry.builder()
                .status(AnimalStatus.Active)
                .changedAt(now)
                .changedBy(userContext.getUserId())
                .changedByUsername(userContext.getUsername())
                .note("Animal registered")
                .build();

        Animal animal = Animal.builder()
                .name(request.name())
                .category(request.category())
                .breed(request.breed())
                .gender(request.gender())
                .ageMonths(request.ageMonths())
                .weight(request.weight())
                .color(request.color())
                .chipId(request.chipId() != null && !request.chipId().isBlank() ? request.chipId() : null)
                .status(AnimalStatus.Active)
                .description(request.description())
                .registeredBy(userContext.getUserId())
                .registeredByUsername(userContext.getUsername())
                .createdAt(now)
                .updatedAt(now)
                .build();

        animal.getStatusHistory().add(initialStatus);

        animal = animalRepository.save(animal);

        // Sync to Weaviate vector store for RAG search
        animalVectorService.upsertAnimal(animal);

        eventPublisher.publishAnimalRegistered(AnimalRegisteredEvent.builder()
                .animalId(animal.getId())
                .name(animal.getName())
                .category(animal.getCategory().name())
                .breed(animal.getBreed())
                .gender(animal.getGender() != null ? animal.getGender().name() : null)
                .ageMonths(animal.getAgeMonths())
                .weight(animal.getWeight())
                .registeredBy(userContext.getUserId())
                .registeredByUsername(userContext.getUsername())
                .timestamp(now)
                .build());

        return AnimalResponse.fromEntity(animal);
    }

    public AnimalResponse getAnimalById(String id) {
        Animal animal = animalRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Animal not found with id: " + id));
        return AnimalResponse.fromEntity(animal);
    }

    public AnimalListResponse getAllAnimals() {
        List<Animal> animals = animalRepository.findAll();
        List<AnimalResponse> responses = animals.stream()
                .map(AnimalResponse::fromEntity)
                .toList();
        return new AnimalListResponse(true, responses, responses.size());
    }

    public AnimalListResponse searchAnimals(String name, AnimalCategory category,
                                            AnimalStatus status, String chipId) {
        Query query = new Query();

        if (name != null && !name.isBlank()) {
            query.addCriteria(Criteria.where("name").regex(name, "i"));
        }
        if (category != null) {
            query.addCriteria(Criteria.where("category").is(category));
        }
        if (status != null) {
            query.addCriteria(Criteria.where("status").is(status));
        }
        if (chipId != null && !chipId.isBlank()) {
            query.addCriteria(Criteria.where("chipId").is(chipId));
        }

        List<Animal> animals = mongoTemplate.find(query, Animal.class);
        List<AnimalResponse> responses = animals.stream()
                .map(AnimalResponse::fromEntity)
                .toList();
        return new AnimalListResponse(true, responses, responses.size());
    }

    public AnimalResponse updateAnimal(String id, UpdateAnimalRequest request) {
        Animal animal = animalRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Animal not found with id: " + id));

        if (request.name() != null) animal.setName(request.name());
        if (request.category() != null) animal.setCategory(request.category());
        if (request.breed() != null) animal.setBreed(request.breed());
        if (request.gender() != null) animal.setGender(request.gender());
        if (request.ageMonths() != null) animal.setAgeMonths(request.ageMonths());
        if (request.weight() != null) animal.setWeight(request.weight());
        if (request.color() != null) animal.setColor(request.color());
        if (request.description() != null) animal.setDescription(request.description());

        if (request.chipId() != null) {
            if (!request.chipId().equals(animal.getChipId())) {
                if (!request.chipId().isBlank() && animalRepository.existsByChipId(request.chipId())) {
                    throw new IllegalArgumentException("An animal with chip ID " + request.chipId() + " already exists");
                }
                animal.setChipId(request.chipId().isBlank() ? null : request.chipId());
            }
        }

        animal.setUpdatedAt(LocalDateTime.now());
        animal = animalRepository.save(animal);

        // Sync updated data to Weaviate
        animalVectorService.upsertAnimal(animal);

        return AnimalResponse.fromEntity(animal);
    }

    public AnimalResponse updateAnimalStatus(String id, UpdateAnimalStatusRequest request) {
        Animal animal = animalRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Animal not found with id: " + id));

        AnimalStatus previousStatus = animal.getStatus();

        StatusHistoryEntry entry = StatusHistoryEntry.builder()
                .status(request.status())
                .changedAt(LocalDateTime.now())
                .changedBy(userContext.getUserId())
                .changedByUsername(userContext.getUsername())
                .note(request.note())
                .build();

        animal.setStatus(request.status());
        animal.getStatusHistory().add(entry);
        animal.setUpdatedAt(LocalDateTime.now());

        animal = animalRepository.save(animal);

        // Sync status change to Weaviate
        animalVectorService.upsertAnimal(animal);

        eventPublisher.publishAnimalStatusChanged(AnimalStatusChangedEvent.builder()
                .animalId(animal.getId())
                .animalName(animal.getName())
                .previousStatus(previousStatus.name())
                .newStatus(request.status().name())
                .changedBy(userContext.getUserId())
                .changedByUsername(userContext.getUsername())
                .note(request.note())
                .timestamp(LocalDateTime.now())
                .build());

        return AnimalResponse.fromEntity(animal);
    }

    public MessageResponse deleteAnimal(String id) {
        Animal animal = animalRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Animal not found with id: " + id));

        imageStorageService.deleteAllForAnimal(id);
        medicalRecordRepository.deleteByAnimalId(id);
        animalRepository.delete(animal);

        // Remove from Weaviate vector store
        animalVectorService.deleteAnimal(id);

        return new MessageResponse(true, "Animal deleted successfully");
    }

    public ImageUploadResponse uploadImages(String animalId, List<MultipartFile> files) {
        Animal animal = animalRepository.findById(animalId)
                .orElseThrow(() -> new ResourceNotFoundException("Animal not found with id: " + animalId));

        int currentCount = animal.getImagePaths() != null ? animal.getImagePaths().size() : 0;
        if (currentCount + files.size() > maxImagesPerAnimal) {
            throw new IllegalArgumentException("Maximum " + maxImagesPerAnimal
                    + " images allowed per animal. Currently has " + currentCount);
        }

        List<String> savedPaths = imageStorageService.saveImages(animalId, files);
        animal.getImagePaths().addAll(savedPaths);
        animal.setUpdatedAt(LocalDateTime.now());
        animalRepository.save(animal);

        return new ImageUploadResponse(true, savedPaths, savedPaths.size() + " image(s) uploaded successfully");
    }

    public MessageResponse deleteImage(String animalId, String imagePath) {
        Animal animal = animalRepository.findById(animalId)
                .orElseThrow(() -> new ResourceNotFoundException("Animal not found with id: " + animalId));

        if (!animal.getImagePaths().remove(imagePath)) {
            throw new ResourceNotFoundException("Image path not found: " + imagePath);
        }

        imageStorageService.deleteImage(imagePath);
        animal.setUpdatedAt(LocalDateTime.now());
        animalRepository.save(animal);

        return new MessageResponse(true, "Image deleted successfully");
    }
}

File: frontend/src/app/services/animal/animal-service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  AnimalResponse,
  AnimalListResponse,
  ImageUploadResponse,
  MessageResponse,
  AnimalCategory,
  AnimalStatus,
  CreateAnimalRequest,
  MedicalRecordResponse,
  MedicalRecordListResponse,
  CreateMedicalRecordRequest,
  MedicalRecordType,
  RagSearchResponse
} from '../../dto/animal';

@Injectable({
  providedIn: 'root'
})
export class AnimalService {

  private readonly animalsUrl = `${environment.apiGatewayUrl}/animals`;
  private readonly medicalUrl = `${environment.apiGatewayUrl}/medical`;

  constructor(private http: HttpClient) {}

  // ── Animal CRUD ──────────────────────────────────────────

  getAllAnimals(): Observable<AnimalListResponse> {
    return this.http.get<AnimalListResponse>(this.animalsUrl);
  }

  searchAnimals(
    name?: string,
    category?: AnimalCategory,
    status?: AnimalStatus,
    chipId?: string
  ): Observable<AnimalListResponse> {
    let params = new HttpParams();
    if (name) params = params.set('name', name);
    if (category) params = params.set('category', category);
    if (status) params = params.set('status', status);
    if (chipId) params = params.set('chipId', chipId);
    return this.http.get<AnimalListResponse>(`${this.animalsUrl}/search`, { params });
  }

  getAnimalById(id: string): Observable<AnimalResponse> {
    return this.http.get<AnimalResponse>(`${this.animalsUrl}/${id}`);
  }

  createAnimal(request: CreateAnimalRequest): Observable<AnimalResponse> {
    return this.http.post<AnimalResponse>(this.animalsUrl, request);
  }

  updateAnimal(id: string, request: Partial<CreateAnimalRequest>): Observable<AnimalResponse> {
    return this.http.put<AnimalResponse>(`${this.animalsUrl}/${id}`, request);
  }

  updateAnimalStatus(id: string, status: AnimalStatus, note?: string): Observable<AnimalResponse> {
    return this.http.patch<AnimalResponse>(`${this.animalsUrl}/${id}/status`, { status, note });
  }

  deleteAnimal(id: string): Observable<MessageResponse> {
    return this.http.delete<MessageResponse>(`${this.animalsUrl}/${id}`);
  }

  // ── Image Management ─────────────────────────────────────

  uploadImages(id: string, files: File[]): Observable<ImageUploadResponse> {
    const formData = new FormData();
    files.forEach(file => formData.append('files', file));
    return this.http.post<ImageUploadResponse>(`${this.animalsUrl}/${id}/images`, formData);
  }

  deleteImage(id: string, imagePath: string): Observable<MessageResponse> {
    const params = new HttpParams().set('path', imagePath);
    return this.http.delete<MessageResponse>(`${this.animalsUrl}/${id}/images`, { params });
  }

  getImageUrl(imagePath: string): string {
    // Images are served through the API gateway
    return `${environment.apiGatewayUrl}/animals/images/${imagePath}`;
  }

  // ── Medical Records ──────────────────────────────────────

  getMedicalRecords(animalId: string, type?: MedicalRecordType): Observable<MedicalRecordListResponse> {
    let params = new HttpParams();
    if (type) params = params.set('type', type);
    return this.http.get<MedicalRecordListResponse>(`${this.medicalUrl}/animal/${animalId}`, { params });
  }

  createMedicalRecord(request: CreateMedicalRecordRequest): Observable<MedicalRecordResponse> {
    return this.http.post<MedicalRecordResponse>(this.medicalUrl, request);
  }

  deleteMedicalRecord(id: string): Observable<MessageResponse> {
    return this.http.delete<MessageResponse>(`${this.medicalUrl}/${id}`);
  }

  // ── RAG Semantic Search ────────────────────────────────────

  ragSearch(query: string, limit: number = 5): Observable<RagSearchResponse> {
    return this.http.post<RagSearchResponse>(`${this.animalsUrl}/rag-search`, { query, limit });
  }
}
